### ISSUE TITLE
Versioning Second Attempt

### ISSUE THREAD

**Issue by oetiker** [qooxdoo/qooxdoo#10161]
If I gather the mood of the core team correctly, we are all on board with semantic versioning. But we are not really clear as to how we can sensibly deal with producing production quality releases and continuing develpment in parallel while keeping everything straight. Another project I know that faces a simpliar problem is Flutter. They use a concept called 'channels' which seems to resolve the issue:

This is the document on how they do it ... and it seems to work as they are actually using this concept for real :)

https://github.com/flutter/flutter/wiki/Flutter-build-release-channels

please discuss.

**Comment by cboulanger** [comment]
As said on Gitter, I favour 

1) a workflow that uses straight x.y.z version numbers with no "-beta" qualification. Code quality should be communicated via channels/tags that are pinned to particular versions/commits. 

2) a versioning model which only Non-BC changes to the developer-facing qooxdoo API proper are considered "breaking". Normally, this excludes the `qx.tool.*` namespace. The reason for this is that a change in `qx.tool.*` in only very rare affects user code. Exception to this is when a change in the tooling would break user code (i.e. backward-incompatible syntax changes).

**Comment by johnspackman** [comment]
@derrell please can you outline your proposal from gitter?

**Comment by oetiker** [comment]
@cboulanger regarding your dislike for extra qualifiers in semantic version numbers which would give the user an instant clue as to what kind of release they are facing ... are there any other projects you can point me to where they are practicing such an approach.

I did point to flutter because they have their whole process mapped out ...

**Comment by cboulanger** [comment]
@oetiker It's basically what is described here: https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/distribution-channels.md . Semantic release is used by a lot of projects: although I couldn't find a list of which exactly, if you search on github for "semantic release" you get a lot of hits.

**Comment by cboulanger** [comment]
In my view, it would remove complexity: with the current system, you always have to decide prospectively what the next version will be to which your are  working towards (i.e., of which the current code is a "beta" of). Whereas with this system, each change is a version change, and you don't care about the versions, because semver and the machine take care of it. That's how NPM works - as long as you don't specify anything, you'll stay on the "stable" branch and don't use any "beta" stuff (regardless of the version, which don't interest you). Only if you decide to take the risk of excplicitly installing `@master`, you know what you're doing an you don't need to be told by the version number. I like that approach, but of course other's might prefer a different one.

**Comment by derrell** [comment]
> @derrell please can you outline your proposal from gitter?

Sure. The main issue I hope to address is that those tracking the bleeding edge branch (which we have agreed to call `master`), of which we seem to have a good many, get the same benefits of semantic release as those following the release branch (which I believe we agreed to call `latest`; whatever @johnspackman had proposed). They should immediately know, based on the version number, if there is a breaking change, new features, etc. What that means is that I would like for the `breaking.feature.fix` version number semantics apply to npm releases built from `master` with each merge, as well as to to npm releases built from `latest` with each merge.

The basic workflow I favor is documented at https://github.com/semantic-release/semantic-release/blob/master/docs/recipes/distribution-channels.md#publishing-on-distribution-channels with the exception that they suggest having the `master` branch publish to the `@latest` channel and the `next` branch publish to the `@next channel`, whereas we would have the `latest` branch publish to the `@latest` channel and the `master` branch publish to some other TBD channel (`@wip` maybe? I don't much care what name we choose.).

I suggest that any user-affecting non-BC change cause a bump to the `breaking` portion of the version number in `@wip`. This means that if we're starting at v6.0.0 with what we have recently released, the first breaking change merged into `master` will bump `@wip` to 8.0.0. If the next merge is a new feature, it would create a new release in `@wip` with the verison number bumpped to 8.1.0. An immediately subsequent merge with a breaking change will create a release with it bumpped to 9.0.0, etc.

That is all in `@wip`, with `@latest remaining at `6.0.0` until we choose to do an "official" release. If we were to do an official release right after the three example changes above, we would merge `master` into `latest`. Those three changes get merged into `latest` -- a breaking change, a bug fix, and another breaking change -- generating implicit version bumps, so that after this merge, we end up building npm release 9.0.0 in the `@latest` channel. There is never any disconnect between the version number that was released in `@wip` and the one release in `@latest`.

The stated goal was to allow those tracking our bleeding edge to get the benefits of semantic release, including awareness through version numbers of breaking changes, while tracking the `@wip` channel. This proposal meets that goal.

There is a consequence to using semantic release exactly as described in the workflow. The consequence is that we don't know what semantic release version number will be the next one we release to `@latest` until we decide to make that release. We therefore can't be talking about the next release by its semantic release version number; rather, we should have a different terminology for talking about subsequent releases to `@latest`. I had suggested _naming_ (vs. numbering) releases to `@latest` for this purpose, and I suggested using names where the first character of the name increments by one letter with each release, so one knows that, for example, the release named `Battlestar` is the one that immediately follows `Andromeda`. Using this terminlogy, the above three example changes, and my example release naming, the `@latest` channel currently has semantic release version 6.0.0 and name `6.0.0` (since we weren't yet using alpha names to describe releases). The next release we do to `@latest` would be semantic release version number `9.0.0` (as described above) and we refer to that release as `Andromeda`. The release after that might be semantic release verison number `9.6.0` (no breaking changes and six new features), and we'd refer to it as the `Battlestar` release. etc.

**Comment by oetiker** [comment]
@derrell so the idea would be to more or less designate a particular 'breaking' release to be an LTS release we could even do that on a schedule ... and the LTS release would continue to get bugfixes. Features would not be backported ... 

This would also mean that what is happening at the moment with master being 'broken' to some extent would not be allowed. THe Breaking 7.0 PR would have to be developped in a branch until it is in a state where the system continues to work after the merge.

**Comment by derrell** [comment]
The point at which we release to `@latest` need not be a breaking change. At any time we want, when we feel that `master`/`@wip` is at a state that we want to officially release it, we just merge it into `latest` which creates a new release in `@latest`. That could be on a schedule, or when we release some fabulous new feature, or whatever we choose.

Regarding backports, if we're releasing often enough, they shouldn't be necessary, unless we wanted to re-release, for example, Andromeda with some fixes because Battlestar had breaking changes that not everyone would want to deal with. I'm not at all convinced it's necessary or that we should be spending time on backports. With our automatic migration handling most of the upgrade process (we hope), I'd rather encourage users to run the most recent code in `@latest`.

Ideally, `master` wouldn't be broken, but since it's bleeding edge, that's not guaranteed. People tracking the `@wip` channel should be prepared for some breakage. If, though, we know that we'll be breaking things by commiting to `master`, we probably should work in a separate branch until its reasonably stable.

**Comment by johnspackman** [comment]
## The Tool Proposal
There are two independent proposals which seem to have become merged - firstly, there is `semantic-release` (the tool) which is simply a mechanism to take the comments in commit messages and use them to automatically generate CHANGELOGs and also to trigger version number bumping - the idea being that, over time, it is easy to forget or miss relevant changes.

We seem to be all quite happy using the semantic versioning kind of numbering, and therefore using `semantic-release` (the tool) seems to be a sensible solution; the previous proposal was just to use the tool and how that tool would formalise the branch names that we use - and with only a few minor additions, this would follow what we have done with 6.x as well as in 5.x before that (and probably earlier versions too).

## The Release Process Proposal
The second proposal (which is largely the subject of this issue) is to change the way that we issue releases and version numbers.

Personally, I’m quite happy with how we have managed versioning and releases in the past and see the disadvantages of the proposed change as:

(1) Every PR which contains a breaking change issues a new major version number; even if there are several PRs which have to happen over time, or if corrective PRs need to be issued, the version is bumped.  If a subsequent PR “unbreaks” a previous breaking change, it’s two version bumps but net zero breaking changes (this happened with the Object ID PR).

(2) Major releases are not just serious and obvious breaking changes, they can include all kinds of minor tweaks and adjustments which will affect the end user (and have a “breaking” effect to some degree for some people).

Under this proposal, the question of “what is a breaking change” becomes much more frequently important and means that it will be debated in every PR; is removing deprecated code a breaking change?  What about a subtle change in the way that layouts work?  Or relocating files?  

At some point, a non-critical bugfix can be seen as a breaking change - or sometimes it is difficult to work out definitively whether the change would be able to “break” someone else’s code.

(3) If we encourage major version numbers as an indication of stability in our beta releases, how do we flag that a feature release or bug fix could cause instability?

For example, adding `qx.Promise` support to events and apply methods was 100% backwards compatible but temporarily very de-stablising.

Having a “beta” track (which is the `master` branch) with a semver numbering system that behaves like “release” (eg without `-beta` or similar suffix and where the major version number changes) could be inferred that it is more stable than it actually is.  

(4) When looking at the public/release track, the current version number would have huge jumps - eg from 6.0 to 9.0; intuitively, upgrades between single digits (eg 6.0 to 7.0) are likely to have *some* issues, but a leap across several version numbers suggests that there are a *lot* of changes.  

Imagine an app written in Qooxdoo v3 and someone wanted to upgrade it to 6.0?  They could expect the changes to be huge (and rightly so); but this proposal would suggest that 6.0 to 9.0 is a single digit upgrade.

### Multiple channels (eg Flutter and others)
Using flutter as an example, they have a `master` branch and eventually push collections of changes into their dev/beta/stable channels; however this issue proposal is that everything always gets pushed automatically and immediately into a beta channel.

Products like flutter with dev/beta/stable channels have a review and merge process for pushing code progressively through the channels - this takes time and human resources that we do not have and could easily be argued to be better spent coding (or working for a living).  Flutter’s [description of their process](https://github.com/flutter/flutter/wiki/Release-process) is too complex for them to describe at the moment.

### Version Naming
The idea of names for launching products is nice, but the need to use them for identifying versions seems to derive from dealing with the issues caused by the major version number increasing rapidly (eg the disadvantages above).

IMHO, 7 > 6 is more intuitive than Ganymede > Ferengi, I am guaranteed to not misspell 6 or 7, and I don’t have to count the alphabet off on my fingers to translate between numbering systems 🙂

I don’t understand whether the version names are proposed for the public releases only (eg Andromeda is 6.0 and Battlestar is 9.0), but if Andomeda, Battlestar, Chekov are contiguous but the versions that they represent are 6.0, 9.0, 11.0 - I would think that is even more confusing (not least because you then have to Google it to switch between version names and numbers).

## The Status Quo
We have historically maintained `master` very well and as reasonably stable platform; it is Git based (as opposed to SVN/CVS type) so it is relatively easy to manage when your own repo needs to step back from the bleeding edge of the Qooxdoo master.

`master` is known to be alpha/beta status and may break - the CHANGELOG (which could be updated automatically by `semantic-release` the tool) lists changes as they are added, and if it’s not explicitly flagged as “breaking” the reader can still decide how “breaking” they are (or what needs to be reviewed).

IMHO the proposal encourages making the major version number a substitute for everything in the CHANGELOG and a substitute for careful review before applying, and at the same time puts a heavier burden on the PR process, and will make major version numbers increasingly meaningless.  

Periodically deciding when to make releases means that we can do it in our own time.

There does not seem to be a major problem to solve (because we’re quite good at keeping `master` stable).  

## Faster Releases
Ideally we should have more frequent releases - 6.0 was a very long time coming, largely because there was a lot of work involved.  

Now that 6.0 is out there, how about if we have regular annual release dates?  Eg 6.0 was Spring 2021, 7.0 could be Spring 2022.

**Comment by oetiker** [comment]
I wonder what it is with breaking changes. From a developer point of view, I love me a library which does not break stuff ... because then I can upgrade without much fuss ... so I think it would be beneficial for the qx if we made it annoying for ourselves to break stuff ... and looking back we have broken very few things, but our processes have not been designed to take that into account, as a feature ... 

And if we do not break things, then fast releases are very simple we could even commit to release every 3 months or something ... as it would be largely an automated process ...

**Comment by cboulanger** [comment]
I think all arguments have been exchanged. The question of whether or not to use semantic release or not is not technical, but philosophical. If you think that major version numbers are important _for humans_, SR is definitely not for you. As I have argued before, I  personally think that version numbers should be for computers, and the number doesn't carry any significance. That's the approach Firefox, Chrome  have taken, with FF being at 86, chrome at 89, and nobody cares. I think that's the future. 

But I certainly do not want to force this opinion on anyone else. In this project, I care deeply about that everybody has fun contributing and can do so the way they like. And since @johnspackman feels strongly about the subject, I don't think we should take a formal vote about adopting SR, but rather find a different solution. 

However, as I have felt responsible for the migration process, there is one thing *I* feel strongly about, and that is that I really think that having a moving target such as `7.0.0-beta` is a bad idea, in particular if a lot of stuff changes during the beta. If I want to have a deterministic migration process, I need to know from where I am migrating. Of course, we can re-introduce versions like `7.0.0-beta.20210323232323`, but that seems ridiculous, because then we might as well have a real version number, the difference being that we are already in v7 land in this case.

So let me re-iterate one more time that I think it makes more sense to me ti change versions with each change, rather than rather than having to determine the version that we are working against, because you never know wether the next version will be "big" or just "breaking". And above I have tried to give a very precise definition of what "breaking" means, so I don't see any major problem determining if a PR is breaking or not. 

But maybe there's a way these different ideas can be reconciled.
